### ISSUE TITLE
sequence to sequence modeling

### ISSUE THREAD

**Issue by sun-peach** [fchollet/keras#3262]
Hi,

I am pretty new to RNN and Keras. So my question may be a little bit naive.

Recently, I am doing some work on sequence-to-sequence problem. But my work is not about the translation or dialog system as previous papers stated. My data is like this: 
1. I have several sequence of feature frames. They have different length
2. Each sequence was collected during one period.
2. Every feature frame has a label.
Now, I want to build up a RNN that when I input a sequence of feature frames, it will output a sequence of labels (one label per feature frame).

I have read some tutorials about training RNN with Keras. However, most of them are talking about segmenting (or padding) the data into sequences with fixed length; each sequence only have one label. Some others are talking about sequence to sequence by using two RNN (one for encoder, the other for decoder). These are different from what I need.

Does anyone have any experience about building up model for my case?

Thank you very much!

**Comment by yurkor** [comment]
Seems like you have sequence of frames to sequence of labels problem, wheres seq in and seq has same length.

You probably want something like this if every frame is discrete enum(like word)

```
model = Sequential()
model.add(Embedding(input_dim=TS, output_dim=100))
for _ in range(10):
   model.add(GRU(64, return_sequences=True))
model.add(TimeDistributed(Dense(PS)))
model.add(Activateion("softmax))         
model.compile(Nadam(clipnorm=10),"sparse_categorical_crossentropy",metrics=["accuracy"]) 
```

if your input is frame is image and then you have to replace Embedding with sequence of Convolution3D or TimeDistributed(Convolution2D)

if you output is image remove model.add(Activateion("softmax))

**Comment by sun-peach** [comment]
@yurkor 
Thank you very much for your help.
It help me a lot. My case takes 14-D feature frame of continuous number as input and the the output is just the label among 4 classes.

So I changed the code a little bit shown as following:
`model = Sequential()
    model.add(Dense(64,input_dim=14))
    model.add(GRU(64, return_sequences=True))
    model.add(TimeDistributed(Dense(4)))
    model.add(Activation("softmax"))
    model.compile(loss='categorical_crossentropy', optimizer='Adagrad')`

I only put one recurrent layer here (as a toy network). Am I making this right?

Thank you once again.

**Comment by sun-peach** [comment]
Also, my data include the data of varied length. If I don't mistake, I should use zero padding and put a mask layer at the input. Is that right?

Should I remove the mask layer when I test the network?

Thank you.

**Comment by yurkor** [comment]
1. you input should accept 3d tensor (batch_size, max_time_steps, 14) so either remove model.add(Dense(64,input_dim=14)) and make GRU you first layer or use model.add(TimeDistributed(Dense(N,input_dim=14)))
2. for varied length you want to align all you input them to maximum size by padding zero. you can do this for all your data or just for mini batch(in this case you can sort input by size). You can try mask layer but it does not work with all layers so be careful.

**Comment by sun-peach** [comment]
@yurkor 
Thank you. I am still a little bit confused about how to prepare my data. If I want to use 10 timesteps, should I make chunks storing 10 feature frames?

Or, may I model them without define fixed timesteps? just send the training sequence and corresponding label into it (let the model take care the time dependency itself)

Thank you. Sorry for the naive question.

**Comment by yurkor** [comment]
if ten is the maximum length of you input example, you want input tensor have shape X.shape == (NUMBER_OF_EXAMPLES, 10, 14)
and output
y.shape == (NUMBER_OF_EXAMPLES,10,4) (one hot encoding for last dimension).

There are utils to make padding job easier http://keras.io/preprocessing/sequence/

**Comment by sun-peach** [comment]
@yurkor 
Thank you.

The code works! However, I want to explore more complicated setting. I would like to add some dense layers before the recurrent layer(3 regular layers then 1 recurrent layer and then the softmax). If this is the case, could I just add Dense() before the GRU layer or the TimeDistributed ones?

Also, the timesteps is a parameters that I need manually set, is there any way that I can get rid of this parameter? I would like the network to take care it automatically.

Thank you.

**Comment by tanmri** [comment]
Suppose I have a sequence of word with integer index with length 30 and out  put is a list of integer index of length 15. And I consider return sequence =True. Is is right to use LSTM model? here is me code and error

Recently I am working on slot fillings using LSTM. I get stuck in a problem.


Suppose I have input of length 30 and out put length 15.

## for simplicity I just consider 1 sentence cases ( though there are 25000 sentence and labels)
Suppose I have two list of integers array each of length 15. i.e a= [1,2,3,54,....,23] of length 15 and another is b=[4,10,23,....,20] of length 15. I have to give input as [a,b] i.e list of 2 arrays or concatenate  two array like as [a+b] which is of length 30 now.

out put is of length 15

so what i wanted to do is I have a sentence  and i want to classify which word goes to which class. there are 127 classes.  So here I have a list of integer index of words(length 15) + another list of integers(length 15). So total length is 30.




print(label)   ### here label is one hot encoded

 [[[0. 0. 0. ... 0. 0. 1.] [0. 0. 0. ... 0. 0. 1.] [0. 0. 0. ... 0. 0. 1.] ... [0. 0. 0. ... 0. 0. 1.] [0. 0. 0. ... 0. 0. 0.] [0. 0. 0. ... 0. 0. 0.]]]



label.shape

Out[1117]: (1, 15, 127)

sent Out[1118]: array([[554, 194, 268, 613, 62, 16, 711, 234, 481, 20, 40, 58, 234, 707, 708, 126, 126, 126, 48, 126, 126, 27, 126, 126, 33, 126, 126, 126, 78, 123]], dtype=int32)

sent.shape

Out[1119]: (1, 30)

 if sent.shape[1] > 1: ## ignore 1 word sentences

        loss = model.train_on_batch(sent, label)    #### here I got the error

this the error Then I got an error InvalidArgumentError (see above for traceback): Incompatible shapes: [1,15,127] vs. [1,30,127]
this is the structure i have used

model = Sequential()
model.add(Embedding(n_vocab,100))     ## n_vocab =710
 model.add(Dropout(0.25))
model.add(LSTM(100,return_sequences=True))
 model.add(Dropout(0.25))
model.add(TimeDistributed(Dense(n_classes, activation='softmax')))
model.compile('rmsprop', 'categorical_crossentropy')

I am using the following tutorial. But when input length and out put length is different then problem arise.

https://chsasank.github.io/spoken-language-understanding.html
